### DISCUSSION TITLE
Migrating from EXE to MSI doesn't uninstall old bootstrapper from Control Panel

### DISCUSSION THREAD

**Discussion by nrcaliendo** [wixtoolset/issues#7951] [Questions]
I'm not a WiX expert, so I apologize if my WiX vernacular is lacking.  I've been tasked with migrating our app's installer from an EXE to an MSI, and I'm running into this problem...

Previous versions of our application installed using an EXE that bundled a custom UI bootstrapper, a .NET Framework 4.8 preqreq package, and the app's MSI.

For the sake of this discussion, let's call these versions 1, 2 and 3.  Starting with version 4, our stakeholders want an MSI install only (no custom UI, no .NET prereq).

The problem is when I upgrade the app from, for example, version 3 to version 4 by running the version 4 MSI, the app seems to upgrade successfully, but I'm left with two entries in Add/Remove Programs (ARP) in Control Panel.  For example:

   _My App 3.0
   My App 4.0_

Likewise, if I uninstall My App 4.0, the _My App 3.0_ ARP entry remains.

I assume the _My App 3.0_ ARP entry is associated with the version 3 bundle, which has its own product upgrade code.  What is the recommended way of ensuring these old ARP entries get removed when upgradig via the new MSI?

**Comment by robmen** [comment]
Easiest way is to use the `MajorUpgrade` element. [Deployment Dojo S1:E4 - Out with the Old. In with the New](https://robmensching.com/deployment-dojo/episodes/s1/e4/out-with-the-old-in-with-the-new/) is the beginning of a few episodes where I explore upgrades.

**Reply by nrcaliendo** [reply]
Thanks for your response.  I watched your video and tried your suggestion, using <MajorUpgrade AllowSameVersionUpgrades="yes"... />.  This did not fix my problem.  The _My App 3.0_ ARP entry still remains after uninstalling version 4.

Again, it seems as if the _My App 3.0_ ARP entry is associated with the version 3 bootstrapper.  In fact, inside my Bundle.wxs file, the bundle has its own ProductUpgradeCode:

<?define ProductUpgradeCode = "20E6FB56-CB34-4E86-AEB5-C3573738C65F"?>

This ProductUpgradeCode is different than the one for the product itself:

<?define ProductUpgradeCode = "93EEDE06-AD76-4786-B488-E387750DB0F4"?>

I'm not sure if these are supposed to be the same?

**Reply by robmen** [reply]
Sorry, my mistake. I was answering a different question, not the one you asked.

There isn't anything built in that will uninstall/upgrade a bundle from an MSI.

**Reply by nrcaliendo** [reply]
OK thanks, I think I'll probably have to write some custom action.
